core_impl/src/main/java/io/opencensus/implcore/tags/SerializationUtils.java

```diff
@@ -91,57 +90,47 @@ private SerializationUtils() {}
 
   // Serializes a TagContext to the on-the-wire format.
   // Encoded tags are of the form: <version_id><encoded_tags>
-  static void serializeBinary(TagContext tags, OutputStream output) throws IOException {
-    final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-    byteArrayOutputStream.write(VERSION_ID);
+  static byte[] serializeBinary(TagContext tags) {
+    // Use a ByteArrayDataOutput to avoid needing to handle IOExceptions.
+    final ByteArrayDataOutput byteArrayDataOutput = ByteStreams.newDataOutput();
+    byteArrayDataOutput.write(VERSION_ID);
 
     // TODO(songya): add support for value types integer and boolean
     for (Iterator<Tag> i = tags.unsafeGetIterator(); i.hasNext(); ) {
       Tag tag = i.next();
-
-      // TODO(sebright): Is there a better way to handle checked exceptions in function objects?
-      IOException ex =
-          tag.match(
-              new Function<TagString, IOException>() {
-                @Override
-                public IOException apply(TagString arg) {
-                  try {
-                    encodeStringTag(arg, byteArrayOutputStream);
-                    return null;
-                  } catch (IOException e) {
-                    return e;
-                  }
-                }
-              },
-              new Function<TagLong, IOException>() {
-                @Override
-                public IOException apply(TagLong arg) {
-                  throw new IllegalArgumentException("long tags are not supported.");
-                }
-              },
-              new Function<TagBoolean, IOException>() {
-                @Override
-                public IOException apply(TagBoolean arg) {
-                  throw new IllegalArgumentException("boolean tags are not supported.");
-                }
-              },
-              Functions.<IOException>throwAssertionError());
-      if (ex != null) {
-        throw ex;
-      }
+      tag.match(
+          new Function<TagString, Void>() {
+            @Override
+            public Void apply(TagString arg) {
+              encodeStringTag(arg, byteArrayDataOutput);
+              return null;
+            }
+          },
+          new Function<TagLong, Void>() {
+            @Override
+            public Void apply(TagLong arg) {
+              throw new IllegalArgumentException("long tags are not supported.");
+            }
+          },
+          new Function<TagBoolean, Void>() {
```

Comment:
seems like we want Functions.booleanUnsupported() to avoid allocating one of these each time.

Reply:
I opened #625 to address this.  I didn't add the constants to the API, though, since they won't be needed once we support boolean and long.